### DISCUSSION TITLE
Raspberry Pi Pico not achieving USB1.1 speeds

### DISCUSSION THREAD

**Discussion by tinkerbeast** [hathach/tinyusb#1134] [Q&A]
I am trying out a slightly modified version of the cdc_dual_ports example where I am constantly pumping data to the USB interface. However I am not able to achieve speeds beyond 240KiB/s. 

I checked if it's a processor throughput issue and that's not it. The processor is pumping data faster than the port is consuming.

I was wondering if there are some settings I can play around with to test the speeds.

**Comment by hathach** [comment]
yeah, all examples are configured to use as minimal RAM as possible to fit into smaller MCUs. You could tune it up https://github.com/micropython/micropython/issues/7479#issuecomment-873623632

**Reply by tinkerbeast** [reply]
I increased CFG_TUD_CDC_EP_BUFSIZE, CFG_TUD_CDC_TX_BUFSIZE, CFG_TUD_CDC_RX_BUFSIZE to 512 as suggested in the reference bug. However I still only get about 0.847 Mb / sec throughput when transmitting from Pico to PC.

My transmit code is as follows:
```
void print_1k(int itf) {  
  for(uint32_t i = 0; i < 1000; i++) {
    uint8_t ch = rand();
    tud_cdc_n_write_char(itf, ch);
  }
  tud_cdc_n_write_flush(itf);
}

void cdc_task() {
  int itf = 0;
  if (tud_cdc_n_available(itf)) {
    print_1k(itf);
  }
}
```
Note that CFG_TUD_CDC is 1.

The receive methodology is same as python code in the reference bug.

**Reply by HiFiPhile** [reply]
You mean megabytes ou megabits ?  0.8MB/s is nearly the maximum of USB full-speed, there is #1064 for CDC high throughput transfer.

**Reply by hathach** [reply]
you really need to do profiling like the micrpython folks does, It is mostly the application, test setup. The example is written so that it runs on most mcu(s) especially one with limited resource. I haven't focus much on throughput, though since MP can get it very closed to the specs, it is probably not an issue with the stack itself. When all the features are implemented and stable enough, I will be more focus on the throughput later on.

**Reply by tinkerbeast** [reply]
@HiFiPhile I mean 0.8 Mega-bit per second.

@hathach It seems to be an issue with `tud_cdc_n_write_char` API. I tried the same code with `tud_cdc_write` API and that gives a throughput close to 6.6 Mega-bit per seconds.

**Reply by hathach** [reply]
that is not the issue with `write_char()`, it is just the issue with **your application didn't make the correct usage of the API**. calling a function 1000 time will always be slower than doing it one, and that is not mentioning the actual usb transferring triggered by the 1000 calls.
